go/arrow/compute/internal/kernels/_lib/base_arithmetic.cc

```diff
@@ -0,0 +1,255 @@
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, int16_t, int16)   \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, int32_t, int32)   \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, int64_t, int64)   \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, uint8_t, uint8)   \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, uint16_t, uint16) \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, uint32_t, uint32) \
+  OP_WITH_OVERFLOW(_func_name, _psnip_op, uint64_t, uint64)
+
+OPS_WITH_OVERFLOW(AddWithOverflow, add)
+OPS_WITH_OVERFLOW(SubtractWithOverflow, sub)
+OPS_WITH_OVERFLOW(MultiplyWithOverflow, mul)
+OPS_WITH_OVERFLOW(DivideWithOverflow, div)
+
+// Corresponds to equivalent ArithmeticOp enum in base_arithmetic.go
+// for passing across which operation to perform. This allows simpler
+// implementation at the cost of having to pass the extra int8 and
+// perform a switch.
+//
+// In cases of small arrays, this is completely negligible. In cases
+// of large arrays, the time saved by using SIMD here is significantly
+// worth the cost.
+enum class optype : int8_t {
+    ADD,
+    ADD_CHECKED,
+    SUB, 
+    SUB_CHECKED,
+};
+
+template <typename T>
+using is_unsigned_integer_value = bool_constant<is_integral_v<T> && is_unsigned_v<T>>;
+
+template <typename T>
+using is_signed_integer_value = bool_constant<is_integral_v<T> && is_signed_v<T>>;
+
+template <typename T, typename R = T>
+using enable_if_signed_integer_t = enable_if_t<is_signed_integer_value<T>::value, R>;
+
+template <typename T, typename R = T>
+using enable_if_unsigned_integer_t = enable_if_t<is_unsigned_integer_value<T>::value, R>;
+
+template <typename T, typename R = T>
+using enable_if_integer_t = enable_if_t<
+    is_signed_integer_value<T>::value || is_unsigned_integer_value<T>::value, R>;
+
+template <typename T, typename R = T>
+using enable_if_floating_t = enable_if_t<is_floating_point_v<T>, R>;
+
+struct Add {
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_floating_t<T> Call(Arg0 left, Arg1 right, bool*) {
+        return left + right;
+    }
+
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_integer_t<T> Call(Arg0 left, Arg1 right, bool*) {
+        return left + right;
+    }
+};
+
+struct Sub {
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_floating_t<T> Call(Arg0 left, Arg1 right, bool*) {
+        return left - right;
+    }
+
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_integer_t<T> Call(Arg0 left, Arg1 right, bool*) {
+        return left - right;
+    }
+};
+
+struct AddChecked {
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_floating_t<T> Call(Arg0 left, Arg1 right, bool*) {
+        return left + right;
+    }
+
+    template <typename T, typename Arg0, typename Arg1>
+    static constexpr enable_if_integer_t<T> Call(Arg0 left, Arg1 right, bool* failure) {
```

Comment:
What about replace SFINA with `if constexpr`? Should be easier to read.

Reply:
that's a good idea, i didn't even know about `if constexpr` until now :) I'll update this

Reply:
updated